### PR TITLE
Dynamic input shapes with sharding

### PR DESCRIPTION
**Pull request by cowanmeg**
Adds support for dynamic shape on sharded tensors. Previously, if a global tensor was sharded we could only support concrete sized inputs. 

- Updates tensor metadata to store logical size as the unsharded tensor size when the tensor is sharded. 
- Sharding tests checks both concrete and symbolic sized input tensors.

### CODE REVIEW

File: csrc/multidevice/utils.cpp [hunk 1]
```diff
@@ -192,4 +192,19 @@ std::set<DeviceIdxType> involvedDevices(Expr* expr) {
...
+    TensorView* tv,
+    c10::IntArrayRef sharded_sizes) {
+  std::vector<int64_t> unsharded_sizes;
+  std::copy(
+      sharded_sizes.begin(),
+      sharded_sizes.end(),
+      std::back_inserter(unsharded_sizes));
+  if (isSharded(tv)) {
+    int num_devices = tv->getDeviceMesh().vector().size();
+    unsharded_sizes[0] = num_devices;
```
Comment (naoyam) [inline csrc/multidevice/utils.cpp:205]: Please leave a note that:

- The outermost dimension is assumed to be the only sharded domain if the tensor is sharded
- The extent of the sharded dimension is assumed to be equal to the number of devices of the mesh

Both are not necessarily true in general.

**Review by naoyam — APPROVED** [review]
LGTM

**Review by samnordmann — APPROVED** [review]

**Comment by cowanmeg** [comment]
!build

**Comment by cowanmeg** [comment]
!build

**Comment by cowanmeg** [comment]
!build

**Pull request merged**